### PR TITLE
Fully simplify conditions from implication checking

### PR DESCRIPTION
**Pull request by ttuegel**
Fixes: #1916 

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/src/Kore/Strategies/Goal.hs [hunk 1]
```diff
@@ -812,11 +817,13 @@ removeDestination lensRulePattern mkState goal =
...
-            let configAndRemoval =
-                    fmap (configuration <*) removal
+            let configAndRemoval = fmap (configuration <*) removal
+                sideCondition =
+                    Pattern.withoutTerm configuration
+                    & SideCondition.fromCondition
             simplifiedRemoval <-
                 simplifyConditionsWithSmt
-                    SideCondition.top
+                    sideCondition
```
Comment (ana-pantilie) [inline kore/src/Kore/Strategies/Goal.hs:826]: Why is the side condition necessary here?

File: kore/src/Kore/Strategies/Goal.hs [hunk 2]
```diff
@@ -979,14 +986,29 @@ removalPatterns
...
+                simplifyEvaluatedMultiPredicate
+                    sideCondition
+                    (MultiAnd.make
+                        [ unifiedConditions
+                        , destinationConditions
+                        ]
+                    )
             let remainderPatterns =
-                    fmap remainderPattern unifiedConfigs
+                    fmap Pattern.fromCondition_ remainderConditions
```
Comment (ana-pantilie) [inline kore/src/Kore/Strategies/Goal.hs:1011]: Ah, so we also needed to first fully simplify the destination and all the resulting conditions together before calling `simplifyEvaluated` for `Exists` and `Not`, right?

File: kore/test/Test/Kore/Strategies/Goal.hs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+                Pattern.withCondition
+                    (Mock.f (mkElemVar Mock.x))
+                    (Condition.fromPredicate
+                        $ makeCeilPredicate Mock.testSort
+                        $ Mock.f (mkElemVar Mock.x)
+                    )
+            destination = Pattern.fromTermLike (mkElemVar Mock.y)
+            existentials = [Mock.y]
+
+        _output <- removalPatterns destination configuration existentials
```
Comment (ana-pantilie) [inline kore/test/Test/Kore/Strategies/Goal.hs:35]: Shouldn't this be compared to the expected result?

File: kore/src/Kore/Strategies/Goal.hs (hunk 1 above)
Comment (ttuegel) [reply kore/src/Kore/Strategies/Goal.hs:826]: This side condition will be used for function evaluation while simplifying the argument, before `simplifyConditionsWithSmt` sends the result to the solver.

File: kore/src/Kore/Strategies/Goal.hs (hunk 2 above)
Comment (ttuegel) [reply kore/src/Kore/Strategies/Goal.hs:1011]: Yes, that's right: the simplifiers for `Exists` and `Not` assume that their arguments are fully simplified, but they don't check. What makes it even harder is that _usually_ the patterns are fully simplified at this point, but in a few cases they need extra function or simplification rules to be applied.

**Pull request merged**